language/bytecode_verifier/src/signature.rs

```diff
@@ -8,140 +8,310 @@ use types::vm_error::{StatusCode, VMStatus};
+                        c, ty, k
+                    )),
+                )
+            }),
+    );
+    errors
 }
 
-/// Field definitions have additional constraints on signatures -- field signatures cannot be
-/// references or mutable references.
-pub(crate) fn check_signature_refs(
-    view: &FieldDefinitionView<'_, CompiledModule>,
-) -> Option<VMStatus> {
-    let type_signature = view.type_signature();
-    let token = type_signature.token();
-    let kind = token.signature_token_kind();
-    match kind {
-        SignatureTokenKind::Reference | SignatureTokenKind::MutableReference => {
-            Some(VMStatus::new(StatusCode::INVALID_FIELD_DEF_REFERENCE))
+/// Checks if the given type is well defined in the given context. No references are permitted.
+fn check_signature_no_refs(
+    context: (&[StructHandle], &[Kind]),
+    ty: &SignatureToken,
+) -> Vec<VMStatus> {
+    use SignatureToken::*;
+
+    let (struct_handles, _) = context;
+
+    match ty {
+        U64 | Bool | String | ByteArray | Address | TypeParameter(_) => vec![],
+        Reference(_) | MutableReference(_) => {
+            // TODO: Prop tests expect us to NOT check the inner types.
+            // Revisit this once we rework prop tests.
+            vec![VMStatus::new(StatusCode::INVALID_SIGNATURE_TOKEN)
+                .with_message("reference not allowed".to_string())]
+        }
+        Struct(idx, type_actuals) => {
+            let sh = &struct_handles[idx.0 as usize];
+            check_generic_instance(context, &sh.type_formals, type_actuals)
         }
-        SignatureTokenKind::Value => None,
     }
 }
 
-/// Check that this token is structurally correct.
-/// In particular, check that the token has a reference only at the top level.
-pub(crate) fn check_structure(token: &SignatureToken) -> Option<VMStatus> {
+/// Checks if the given type is well defined in the given context. References are only permitted
+/// at the top level.
+fn check_signature(context: (&[StructHandle], &[Kind]), ty: &SignatureToken) -> Vec<VMStatus> {
     use SignatureToken::*;
 
-    let inner_token_opt = match token {
-        Reference(token) => Some(token),
-        MutableReference(token) => Some(token),
-        Bool | U64 | String | ByteArray | Address | Struct(_, _) | TypeParameter(_) => None,
-    };
-    if let Some(inner_token) = inner_token_opt {
-        if inner_token.is_reference() {
-            let msg = format!(
-                "Invalid token {:#?} of kind {} with inner token {}",
-                token.clone(),
-                token.signature_token_kind(),
-                inner_token.signature_token_kind(),
-            );
-            return Some(VMStatus::new(StatusCode::INVALID_SIGNATURE_TOKEN).with_message(msg));
-        }
+    match ty {
+        Reference(inner) | MutableReference(inner) => check_signature_no_refs(context, inner),
+        _ => check_signature_no_refs(context, ty),
```

Comment:
This seems like a lot to review. It might be easier to make sure that we have tests to cover all of the relevant behaviors if you either split the PR or your tests along the conceptual divides mentioned in your summary:

```
References are only allowed in locals.
Nested references are never allowed.
The number of type formals & actuals must match for all instantiations.
The type actuals must satisfy the kind constraints declared in the formals.
```